### DISCUSSION TITLE
trouble using OCRDataset

### DISCUSSION THREAD

**Discussion by kghezelbash** [hezarai/hezar#151] [Q&A] [answered]
Hi again, I am trying to a make dataset with OCRDataset class of this project. In the OCRDatasetConfig, the 'path' could be a path to a csv file? or I should convert my data to another format? If it does not support csv format why it gets the column names?

**Comment by arxyzan** [comment, marked as answer]
Hello @kghezelbash,
To make it more clear, if you want to have your own class to be able to train your model using Hezar, you have to provide a regular PyTorch Dataset subclass. Hezar has its own dataset classes for casual tasks like OCR, image captioning, text classification, etc. There is no force to only use those classes, they're just there to make it easier for you. If your dataset is so different or need a lot of customizations, you can easily write your own dataset class. 

You can use this template as an example:
```python
from hezar.models import CRNNImage2TextConfig, CRNNImage2Text
from hezar.preprocessors import ImageProcessor
from hezar.trainer import Trainer, TrainerConfig

from hezar.data import OCRDataset, OCRDatasetConfig


class PersianOCRDataset(OCRDataset):
    def __init__(self, config: OCRDatasetConfig, split=None, **kwargs):
        super().__init__(config=config, split=split, **kwargs)

    def _load(self, split=None):
        # Load a dataframe here and make sure the split is fetched
        data = pd.read_csv(self.config.path)
        # preprocess if needed
        return data

    def __getitem__(self, index):
        path, text = self.data.iloc[index].values()
        pixel_values = self.image_processor(path, return_tensors="pt")["pixel_values"][0]
        labels = self._text_to_tensor(text)
        inputs = {
            "pixel_values": pixel_values,
            "labels": labels,
        }
        return inputs


dataset_config = OCRDatasetConfig(
    path="path/to/csv",
    text_split_type="char_split",
    text_column="label",
    images_paths_column="image_path",
    reverse_digits=True,
)

train_dataset = PersianOCRDataset(dataset_config, split="train")
eval_dataset = PersianOCRDataset(dataset_config, split="test")

model = CRNNImage2Text(
    CRNNImage2TextConfig(
        id2label=train_dataset.config.id2label,
        map2seq_in_dim=1024,
        map2seq_out_dim=96
    )
)
preprocessor = ImageProcessor(train_dataset.config.image_processor_config)

train_config = TrainerConfig(
    output_dir="crnn-plate-fa-v1",
    task="image2text",
    device="cuda",
    batch_size=8,
    num_epochs=20,
    metrics=["cer"],
    metric_for_best_model="cer"
)

trainer = Trainer(
    config=train_config,
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    data_collator=train_dataset.data_collator,
    preprocessor=preprocessor,
)
trainer.train()
```

**Comment by kghezelbash** [comment]
Hi again, if we want to finetune the OCR model with text_split_type='tokenize' what id2label we should use?

**Comment by arxyzan** [comment]
@kghezelbash Hi. the tokenize type is only applicable for transformer models like TrOCR that need a tokenizer so that id2label is not necessary. Generally, we do not recommend using that method since it was never tested.

**Reply by kghezelbash** [reply]
Thank you so much for your time and consideration, and your answer. Where in the training config I should change the pre-trained model to TrOCR?

**Comment by arxyzan** [comment]
@kghezelbash Appreciate it. Thanks.
Why would you want to use TrOCR? Since this model requires a lot of labeled data (dozens of millions of samples!) and still won't even beat models like CRNN. Plus, as I mentioned we never tested TrOCR for training in Hezar since we trained it using Transformers beforehand and ported it to Hezar so that only inference was tested thoroughly.

**Reply by kghezelbash** [reply]
I tried to finetune the CRNN model on a big dataset of Persian words written with BYekan, but I was not successful. After fine-tuning the model predicts "" for all the words, so I thought maybe I should change the pre-trained model.

**Reply by arxyzan** [reply]
Actually, this is rare for CRNN to have such corrupt behavior. I think there might be a bug in your training script or Hezar's Trainer. Can you paste or send me your full training script?

**Reply by kghezelbash** [reply]
Of course. In this code "IDDATA2.csv" is a csv containing a path and label for each image. Here is my training script:

import csv
import os
from hezar.data.datasets.ocr_dataset import TextSplitType
from hezar.constants import TaskType
from hezar.data.datasets.ocr_dataset import OCRDatasetConfig
from hezar.preprocessors import ImageProcessor
from hezar.trainer import Trainer, TrainerConfig
import pandas as pd
from hezar.models import CRNNImage2TextConfig, CRNNImage2Text

from hezar.data import OCRDataset, OCRDatasetConfig
from hezar.preprocessors.image_processor import ImageProcessorConfig
from tqdm import tqdm

csv_file = "IDDATA2.csv"

fa_characters = [
    "", "آ", "ا", "ب", "پ", "ت", "ث", "ج", "چ", "ح", "خ", "د", "ذ", "ر", "ز", "ژ", "س", "ش",
    "ص", "ض", "ط", "ظ", "ع", "غ", "ف", "ق", "ک", "گ", "ل", "م", "ن", "و", "ه", "ی", " " ,  "ي"
]
fa_numbers = ["۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹", "۰"]
fa_special_characters = ["ء", "ؤ", "ئ", "أ", "ّ",  'ٓ', 'ٕ', "ٔ", "\u200c" , "j", "p", "g"]
fa_symbols = ["/", "(", ")", "+", "-", ":", "،", "!", ".", "؛", "=", "%", "؟"]
en_numbers = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"]
all_characters = fa_characters + fa_numbers + fa_special_characters + fa_symbols + en_numbers

ID2LABEL = dict(enumerate(all_characters))

image_processor_config = ImageProcessorConfig(
    mean=[0.5],  # Example mean values for normalization
    std=[0.5],   # Example standard deviation values for normalization
    rescale=1.0,           # Example rescaling factor
    resample=2,            # Example resampling filter (2 for BICUBIC)
    size=(224, 224),       # Example target image size (width, height)
    mirror=False,           # Example mirror augmentation
    gray_scale=True       # Example grayscale conversion
)

class PersianOCRDataset(OCRDataset):
    def __init__(self, config: OCRDatasetConfig, split=None, **kwargs):
        super().__init__(config=config, split=split, **kwargs)

    def _load(self, split=None):
        data = pd.read_csv(self.config.path)
        return data

    def __getitem__(self, index):
        path = self.data.iloc[index][0]
        text = self.data.iloc[index][1]
        pixel_values = self.image_processor(path, return_tensors="pt")["pixel_values"][0]
        labels = self._text_to_tensor(text)
        inputs = {
            "pixel_values": pixel_values,
            "labels": labels,
        }
        return inputs


dataset_config = OCRDatasetConfig(
    path=csv_file,  
    text_split_type='char_split', 
    id2label=ID2LABEL,  
    text_column="text",
    images_paths_column="image_path",
    max_length=100,  
    invalid_characters=[],  
    reverse_text=False, 
    reverse_digits=False, 
    image_processor_config=image_processor_config  
)


train_dataset = PersianOCRDataset(dataset_config, split="train")
eval_dataset = PersianOCRDataset(dataset_config, split="test")

model = CRNNImage2Text(
    CRNNImage2TextConfig(
        id2label=train_dataset.config.id2label,
        map2seq_in_dim=7168,
        map2seq_out_dim=96
    )
)

preprocessor = ImageProcessor(train_dataset.config.image_processor_config)

train_config = TrainerConfig(
    output_dir="crnn-plate-fa-v5",
    task="image2text",
    device="CPU",
    batch_size=10,
    num_epochs=10,
    metrics=["cer"],
    metric_for_best_model="cer"
)

trainer = Trainer(
    config=train_config,
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    data_collator=train_dataset.data_collator,
    preprocessor=preprocessor,
)

trainer.train()

**Reply by arxyzan** [reply]
Sorry to ask, but can you wrap the code inside a python block? because this is really hard to read if not wrapped in code block. 
like below:

![image](https://github.com/hezarai/hezar/assets/38841793/fc140781-7155-4c3c-8e40-61bc1cff5aa4)

**Reply by arxyzan** [reply]
I think you used "quotation mark" instead of "back quote". That's why it still shows up incorrectly. You can also use the `Preview` button on top of the text box to see how your text would show up for me.

**Comment by kghezelbash** [comment]
Thank you again =)). 
```python
import csv
import os
from hezar.data.datasets.ocr_dataset import TextSplitType
from hezar.constants import TaskType
from hezar.data.datasets.ocr_dataset import OCRDatasetConfig
from hezar.preprocessors import ImageProcessor
from hezar.trainer import Trainer, TrainerConfig
import pandas as pd
from hezar.models import CRNNImage2TextConfig, CRNNImage2Text

from hezar.data import OCRDataset, OCRDatasetConfig
from hezar.preprocessors.image_processor import ImageProcessorConfig
from tqdm import tqdm

csv_file = "IDDATA2.csv"

fa_characters = [
    "", "آ", "ا", "ب", "پ", "ت", "ث", "ج", "چ", "ح", "خ", "د", "ذ", "ر", "ز", "ژ", "س", "ش",
    "ص", "ض", "ط", "ظ", "ع", "غ", "ف", "ق", "ک", "گ", "ل", "م", "ن", "و", "ه", "ی", " " ,  "ي"
]
fa_numbers = ["۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹", "۰"]
fa_special_characters = ["ء", "ؤ", "ئ", "أ", "ّ",  'ٓ', 'ٕ', "ٔ", "\u200c" , "j", "p", "g"]
fa_symbols = ["/", "(", ")", "+", "-", ":", "،", "!", ".", "؛", "=", "%", "؟"]
en_numbers = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"]
all_characters = fa_characters + fa_numbers + fa_special_characters + fa_symbols + en_numbers

ID2LABEL = dict(enumerate(all_characters))

image_processor_config = ImageProcessorConfig(
    mean=[0.5],  # Example mean values for normalization
    std=[0.5],   # Example standard deviation values for normalization
    rescale=1.0,           # Example rescaling factor
    resample=2,            # Example resampling filter (2 for BICUBIC)
    size=(224, 224),       # Example target image size (width, height)
    mirror=False,           # Example mirror augmentation
    gray_scale=True       # Example grayscale conversion
)

class PersianOCRDataset(OCRDataset):
    def __init__(self, config: OCRDatasetConfig, split=None, **kwargs):
        super().__init__(config=config, split=split, **kwargs)

    def _load(self, split=None):
        data = pd.read_csv(self.config.path)
        return data

    def __getitem__(self, index):
        path = self.data.iloc[index][0]
        text = self.data.iloc[index][1]
        pixel_values = self.image_processor(path, return_tensors="pt")["pixel_values"][0]
        labels = self._text_to_tensor(text)
        inputs = {
            "pixel_values": pixel_values,
            "labels": labels,
        }
        return inputs


dataset_config = OCRDatasetConfig(
    path=csv_file,  
    text_split_type='char_split',  
    id2label=ID2LABEL, 
    text_column="text",
    images_paths_column="image_path",
    max_length=100,  
    invalid_characters=[],  
    reverse_text=False,   
    reverse_digits=False,   
    image_processor_config=image_processor_config  
)


train_dataset = PersianOCRDataset(dataset_config, split="train")
eval_dataset = PersianOCRDataset(dataset_config, split="test")

model = CRNNImage2Text(
    CRNNImage2TextConfig(
        id2label=train_dataset.config.id2label,
        map2seq_in_dim=7168,
        map2seq_out_dim=96
    )
)

preprocessor = ImageProcessor(train_dataset.config.image_processor_config)

train_config = TrainerConfig(
    output_dir="crnn-plate-fa-v1",
    task="image2text",
    device="CPU",
    batch_size=10,
    num_epochs=1,
    metrics=["cer"],
    metric_for_best_model="cer"
)

trainer = Trainer(
    config=train_config,
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    data_collator=train_dataset.data_collator,
    preprocessor=preprocessor,
)

trainer.train()
```

**Reply by arxyzan** [reply]
Nice!
Why didn't you also set a pretrained weight for this model? Do you actually want to train the model from scratch? Even if your labels are not the same with the pretrained model, I recommend you set the `init_weights_from` in the trainer's config or use a pretrained model totally.

My recommendation is to use our latest CRNN model as the base model. If your dataset needs extra characters you can append the alphabet too. Based on your code I think the following would do the job. 

But before that, I have to mention that your script has some flaws actually, 
- First you did not reverse text in the config, this is crucial since CRNN base model was trained on reveresed labels since the model processes the image from left to right, so we needed to reverse the image for Persian, but for scenarios like ALPR where the text must be read from left to right (unlike regular Persian text) so we should either not hflip the image (which would be hard for the base model since it was trained on reversed version of RTL characters) or let it hflip the image but also reverse the text too to achieve faster convergence
- Secondly, it's better to append extra labels to the end of the labels list so that the model does not lose previous knowledge of other characters when finetuning. 
- Thirdly, your image processor configs were wrong. Use the ones in the pretrained models' configs. 

```python
import pandas as pd

from hezar.trainer import Trainer, TrainerConfig
from hezar.models import Model
from hezar.data import OCRDataset, OCRDatasetConfig
from hezar.preprocessors import ImageProcessorConfig

csv_file = "IDDATA2.csv"

fa_characters = [
    "", "آ", "ا", "ب", "پ", "ت", "ث", "ج", "چ", "ح", "خ", "د", "ذ", "ر", "ز", "ژ", "س", "ش",
    "ص", "ض", "ط", "ظ", "ع", "غ", "ف", "ق", "ک", "گ", "ل", "م", "ن", "و", "ه", "ی", " ", "ي"
]
fa_numbers = ["۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹", "۰"]
fa_special_characters = ["ء", "ؤ", "ئ", "أ", "ّ", 'ٓ', 'ٕ', "ٔ"]
fa_symbols = ["/", "(", ")", "+", "-", ":", "،", "!", ".", "؛", "=", "%", "؟"]
en_numbers = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"]
additional_characters = ["\u200c", "j", "p", "g"]  # Append extra characters to the end for faster convergence

all_characters = fa_characters + fa_numbers + fa_special_characters + fa_symbols + en_numbers, additional_characters
ID2LABEL = dict(enumerate(all_characters))


class PersianOCRDataset(OCRDataset):
    def __init__(self, config: OCRDatasetConfig, split=None, **kwargs):
        super().__init__(config=config, split=split, **kwargs)

    def _load(self, split=None):
        data = pd.read_csv(self.config.path)
        return data

    def __getitem__(self, index):
        path = self.data.iloc[index][0]
        text = self.data.iloc[index][1]
        pixel_values = self.image_processor(path, return_tensors="pt")["pixel_values"][0]
        labels = self._text_to_tensor(text)
        inputs = {
            "pixel_values": pixel_values,
            "labels": labels,
        }
        return inputs


if __name__ == '__main__':
    image_processor_config = ImageProcessorConfig(
        mean=[0.6595],
        std=[0.1501],
        rescale=1 / 255,
        size=(384, 32),
        mirror=True,
        gray_scale=True,
    )

    dataset_config = OCRDatasetConfig(
        path=csv_file,
        text_split_type='char_split',
        id2label=ID2LABEL,
        text_column="text",
        images_paths_column="image_path",
        max_length=48,
        invalid_characters=[],
        reverse_digits=True,
        reverse_text=True,
        image_processor_config=image_processor_config
    )

    train_dataset = PersianOCRDataset(dataset_config, split="train")
    eval_dataset = PersianOCRDataset(dataset_config, split="test")

    model = Model.load("hezarai/crnn-fa-printed-96-long", id2label=ID2LABEL)

    train_config = TrainerConfig(
        output_dir="crnn-plate-fa-v1",
        task="image2text",
        device="cpu",
        batch_size=10,
        num_epochs=1,
        metrics=["cer"],
        metric_for_best_model="cer"
    )

    trainer = Trainer(
        config=train_config,
        model=model,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        data_collator=train_dataset.data_collator,
    )

    trainer.train()

```

**Reply by kghezelbash** [reply]
Thank you for your accurate reply.
I applied what you said, but my problem resisted. Is there any possibility that I am using the fine-tuned model wrong? I am using trainer.model after fine-tuning as my new model for prediction.

**Reply by arxyzan** [reply]
You're welcome :))
Can you please paste the output of the Trainer here so that I can see how the loss is going? Also, what is the final value of the `CER` when evaluating the model?

**Reply by arxyzan** [reply]
Note that the final model you have to use can be easily and safely loaded from the checkpoints since those files contain all the necessary configs so that you load the model with the correct properties.
```python
...
model = Model.load("crnn-plate-fa-v1/checkpoints/1")  # Change this path to the folder where your last checkpoint is saved
outputs = model.predict("example_image.jpg")
print(outputs)
```

**Reply by kghezelbash** [reply]
This is the output of the trainer:
******************** Training Info ********************

  Output Directory: `crnn-plate-fa-v1`
  Task: `image2text`
  Model: `CRNNImage2Text`
  Init Weights: `hezarai/crnn-fa-printed-96-long`
  Device(s): `cuda`
  Batch Size: `10`
  Epochs: `5`
  Training Dataset: `PersianOCRDataset(path=/content/drive/MyDrive/IDDATA2.csv['train'], size=903)`
  Evaluation Dataset: `PersianOCRDataset(path=/content/drive/MyDrive/IDDATA2.csv['test'], size=903)`
  Optimizer: `adam`
  Scheduler: `None`
  Initial Learning Rate: `2e-05`
  Learning Rate Decay: `0.0`
  Number of Parameters: `9306418`
  Number of Trainable Parameters: `9306418`
  Mixed Precision: `Full (fp32)`
  Metrics: `['cer']`
  Checkpoints Path: `crnn-plate-fa-v1/checkpoints`
  Logs Path: `crnn-plate-fa-v1/logs/Apr02_08-31-44_66e71ab5dbd0`

*******************************************************


Epoch: 1/5      100%|######################################################################| 90/90 [00:08<00:00, 10.28batch/s, loss=13.7]
Evaluating...     0%|                                                                      | 0/90 [00:00<?, ?batch/s]/usr/local/lib/python3.10/dist-packages/hezar/models/image2text/crnn/crnn_image2text.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(torch.zeros(generated_ids.shape))
Evaluating...   100%|######################################################################| 90/90 [00:08<00:00, 10.07batch/s, cer=1, loss=4.42]

Epoch: 2/5      100%|######################################################################| 90/90 [00:08<00:00, 10.80batch/s, loss=4.12]
Evaluating...     0%|                                                                      | 0/90 [00:00<?, ?batch/s]/usr/local/lib/python3.10/dist-packages/hezar/models/image2text/crnn/crnn_image2text.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(torch.zeros(generated_ids.shape))
Evaluating...   100%|######################################################################| 90/90 [00:08<00:00, 10.50batch/s, cer=1, loss=3.88]

Epoch: 3/5      100%|######################################################################| 90/90 [00:08<00:00, 10.53batch/s, loss=3.72]
Evaluating...     0%|                                                                      | 0/90 [00:00<?, ?batch/s]/usr/local/lib/python3.10/dist-packages/hezar/models/image2text/crnn/crnn_image2text.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(torch.zeros(generated_ids.shape))
Evaluating...   100%|######################################################################| 90/90 [00:08<00:00, 10.58batch/s, cer=1, loss=3.57]

Epoch: 4/5      100%|######################################################################| 90/90 [00:08<00:00, 10.24batch/s, loss=3.47]
Evaluating...     0%|                                                                      | 0/90 [00:00<?, ?batch/s]/usr/local/lib/python3.10/dist-packages/hezar/models/image2text/crnn/crnn_image2text.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(torch.zeros(generated_ids.shape))
Evaluating...   100%|######################################################################| 90/90 [00:08<00:00, 10.56batch/s, cer=1, loss=3.37]

Epoch: 5/5      100%|######################################################################| 90/90 [00:08<00:00, 10.07batch/s, loss=3.3]
Evaluating...     0%|                                                                      | 0/90 [00:00<?, ?batch/s]/usr/local/lib/python3.10/dist-packages/hezar/models/image2text/crnn/crnn_image2text.py:86: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  scores = torch.tensor(torch.zeros(generated_ids.shape))
Evaluating...   100%|######################################################################| 90/90 [00:08<00:00, 10.06batch/s, cer=1, loss=3.23]
Hezar (INFO): Training done!

**Reply by arxyzan** [reply]
Hmm, this seems like an issue in the training. The loss is not converging as it should and the CER score is 1 which means the whole prediction is wrong. The only way to figure this out is that I have access to a small portion (100 samples at least) of your data so that I can run this script to be able to debug it. If you're ok with that, you can give me access or send the data to me.

**Reply by kghezelbash** [reply]
Sure, how can I share it with you?

**Reply by arxyzan** [reply]
Google Drive link or Hugging Face Datasets is okay for me.

**Reply by kghezelbash** [reply]
https://drive.google.com/drive/folders/1zDprx2DHb9vdkbETYjvb8H3awSnI09To?usp=sharing

**Reply by arxyzan** [reply]
Alright, now I see, your dataset is not ALPR at all, it's just standard Persian OCR. 
First off, the images seem pretty easy to recognize and I tested a couple of them and the vanilla CRNN model at [hezarai/crnn-fa-printed-96-long](https://huggingface.co/hezarai/crnn-fa-printed-96-long) can actually recognize them pretty accurately. (I ran a test on 20 samples from your data and not a single one was recognized incorrectly).

But anyway, if you still want to train a model on this data here's the possible working code for that:
```python
import pandas as pd

from hezar.trainer import Trainer, TrainerConfig
from hezar.models import Model
from hezar.data import OCRDataset, OCRDatasetConfig
from hezar.preprocessors import ImageProcessorConfig

csv_file = "IDDATA2.csv"

fa_characters = [
    "", "آ", "ا", "ب", "پ", "ت", "ث", "ج", "چ", "ح", "خ", "د", "ذ", "ر", "ز", "ژ", "س", "ش",
    "ص", "ض", "ط", "ظ", "ع", "غ", "ف", "ق", "ک", "گ", "ل", "م", "ن", "و", "ه", "ی", " ", "ي"
]
fa_numbers = ["۱", "۲", "۳", "۴", "۵", "۶", "۷", "۸", "۹", "۰"]
fa_special_characters = ["ء", "ؤ", "ئ", "أ", "ّ", 'ٓ', 'ٕ', "ٔ"]
fa_symbols = ["/", "(", ")", "+", "-", ":", "،", "!", ".", "؛", "=", "%", "؟"]
en_numbers = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"]
additional_characters = ["\u200c", "j", "p", "g"]  # Append extra characters to the end for faster convergence

all_characters = fa_characters + fa_numbers + fa_special_characters + fa_symbols + en_numbers, additional_characters
ID2LABEL = dict(enumerate(all_characters))


class PersianOCRDataset(OCRDataset):
    def __init__(self, config: OCRDatasetConfig, split=None, **kwargs):
        super().__init__(config=config, split=split, **kwargs)

    def _load(self, split=None):
        data = pd.read_csv(self.config.path)
        return data

    def __getitem__(self, index):
        path = self.data.iloc[index][0]
        text = self.data.iloc[index][1]
        pixel_values = self.image_processor(path, return_tensors="pt")["pixel_values"][0]
        labels = self._text_to_tensor(text)
        inputs = {
            "pixel_values": pixel_values,
            "labels": labels,
        }
        return inputs


if __name__ == '__main__':
    image_processor_config = ImageProcessorConfig(
        mean=[0.6595],
        std=[0.1501],
        rescale=1 / 255,
        size=(384, 32),
        mirror=True,
        gray_scale=True,
    )

    dataset_config = OCRDatasetConfig(
        path=csv_file,
        text_split_type='char_split',
        id2label=ID2LABEL,
        text_column="text",
        images_paths_column="image_path",
        max_length=48,
        invalid_characters=[],
        reverse_digits=True,
        image_processor_config=image_processor_config
    )

    train_dataset = PersianOCRDataset(dataset_config, split="train")
    eval_dataset = PersianOCRDataset(dataset_config, split="test")

    model = Model.load("hezarai/crnn-fa-printed-96-long", id2label=ID2LABEL)

    train_config = TrainerConfig(
        output_dir="crnn-fa-custom",
        task="image2text",
        device="cpu",
        batch_size=16,
        num_epochs=20,
        metrics=["cer"],
        metric_for_best_model="cer"
    )

    trainer = Trainer(
        config=train_config,
        model=model,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        data_collator=train_dataset.data_collator,
    )

    trainer.train()
```

**Reply by kghezelbash** [reply]
You are right. the model works perfectly on this dataset but it has a little problem when the images become blurred, I want to fine-tune the model on a dataset of this font to read those blurred images perfectly too.

**Reply by arxyzan** [reply]
I see. Let me know if you encounter any other problems. Good luck.

**Reply by kghezelbash** [reply]
Thank you so much for your help, it is working now =)))))))))))))

**Reply by arxyzan** [reply]
So happy to hear that! 🍻🍻
